internal/compiler/passes/embed_images.rs

```diff
@@ -10,27 +10,57 @@ use crate::EmbedResourcesKind;
 use image::GenericImageView;
 use std::cell::RefCell;
 use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
 use std::rc::Rc;
 
-pub fn embed_images(
+pub async fn embed_images(
     component: &Rc<Component>,
     embed_files: EmbedResourcesKind,
     scale_factor: f64,
+    resource_url_mapper: &Option<Rc<dyn Fn(&str) -> Pin<Box<dyn Future<Output = Option<String>>>>>>,
     diag: &mut BuildDiagnostics,
 ) {
     let global_embedded_resources = &component.embedded_file_resources;
 
-    for component in component
+    let all_components = component
         .used_types
         .borrow()
         .sub_components
         .iter()
         .chain(component.used_types.borrow().globals.iter())
         .chain(std::iter::once(component))
-    {
+        .cloned()
+        .collect::<Vec<_>>();
+
+    // Collect URLs (sync!):
+    let urls = {
```

Comment:
You only need to collect the Urls if we need to map them  (resource_url_mapper.is_some())